app/src/main/java/uk/ac/ebi/atlas/search/species/SpeciesSearchDao.java

```diff
@@ -0,0 +1,122 @@
+package uk.ac.ebi.atlas.search.species;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.solr.client.solrj.SolrQuery;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+import uk.ac.ebi.atlas.solr.cloud.SolrCloudCollectionProxyFactory;
+import uk.ac.ebi.atlas.solr.cloud.TupleStreamer;
+import uk.ac.ebi.atlas.solr.cloud.collections.BioentitiesCollectionProxy;
+import uk.ac.ebi.atlas.solr.cloud.collections.Gene2ExperimentCollectionProxy;
+import uk.ac.ebi.atlas.solr.cloud.search.SolrQueryBuilder;
+import uk.ac.ebi.atlas.solr.cloud.search.streamingexpressions.decorator.InnerJoinStreamBuilder;
+import uk.ac.ebi.atlas.solr.cloud.search.streamingexpressions.decorator.SelectStreamBuilder;
+import uk.ac.ebi.atlas.solr.cloud.search.streamingexpressions.decorator.UniqueStreamBuilder;
+import uk.ac.ebi.atlas.solr.cloud.search.streamingexpressions.source.SearchStreamBuilder;
+
+import java.util.Optional;
+
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static uk.ac.ebi.atlas.solr.cloud.collections.BioentitiesCollectionProxy.BIOENTITY_IDENTIFIER;
+import static uk.ac.ebi.atlas.solr.cloud.collections.BioentitiesCollectionProxy.BIOENTITY_IDENTIFIER_DV;
+import static uk.ac.ebi.atlas.solr.cloud.collections.BioentitiesCollectionProxy.PROPERTY_NAME;
+import static uk.ac.ebi.atlas.solr.cloud.collections.BioentitiesCollectionProxy.PROPERTY_VALUE;
+import static uk.ac.ebi.atlas.solr.cloud.collections.BioentitiesCollectionProxy.SPECIES;
+import static uk.ac.ebi.atlas.solr.cloud.collections.BioentitiesCollectionProxy.SPECIES_DV;
+
+@Component
+public class SpeciesSearchDao {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(SpeciesSearchDao.class);
+
+    private final BioentitiesCollectionProxy bioentitiesCollectionProxy;
+    private final Gene2ExperimentCollectionProxy gene2ExperimentCollectionProxy;
+
+    public SpeciesSearchDao(SolrCloudCollectionProxyFactory collectionProxyFactory) {
+        bioentitiesCollectionProxy = collectionProxyFactory.create(BioentitiesCollectionProxy.class);
+        gene2ExperimentCollectionProxy = collectionProxyFactory.create(Gene2ExperimentCollectionProxy.class);
+    }
+
+    public Optional<ImmutableSet<String>> searchSpecies(String searchText, String category) {
```

Comment:
The 1st one returns an empty `Optional`.
The 2nd one returns an `Optional` that contains an empty `ImmutableSet`. A `Set` that has no elements.

Reply:
I was referring what does it mean from a semantics point of view (i.e. what it’s telling you in each case, such as not found or the query not being valid, etc.).

Reply:
OK, I got you now. Sorry.
The 1st could mean that we have some issues processing the request, because if we could process the request successfully but we got an empty result set, that's when we response with `Optional.of(ImmutableSet.of())`.
Is that what you meant?